test/syncenginetestutils.h

```diff
@@ -0,0 +1,658 @@
+            auto &dest = this->children[source.name] = source;
+            dest.parentPath = p;
+        }
+    }
+
+    void remove(const QString &relativePath) override {
+        const QStringList pathComponents = relativePath.split('/', QString::SkipEmptyParts);
+        FileInfo *parent = findInvalidatingEtags(pathComponents.mid(0, pathComponents.size() - 1));
+        Q_ASSERT(parent);
+        parent->children.erase(std::find_if(parent->children.begin(), parent->children.end(),
+                [&pathComponents](const FileInfo &fi){ return fi.name == pathComponents.last(); }));
+    }
+    void insert(const QString &relativePath, qint64 size = 64, char contentChar = 'W') override {
+        create(relativePath, size, contentChar);
+    }
+    void setContents(const QString &relativePath, char contentChar) override {
+        FileInfo *file = findInvalidatingEtags(relativePath.split('/', QString::SkipEmptyParts));
+        Q_ASSERT(file);
+        file->contentChar = contentChar;
+    }
+    void appendByte(const QString &relativePath) override {
+        FileInfo *file = findInvalidatingEtags(relativePath.split('/', QString::SkipEmptyParts));
+        Q_ASSERT(file);
+        file->size += 1;
+    }
+    void mkdir(const QString &relativePath) override {
+        createDir(relativePath);
+    }
+    FileInfo *find(const QString &relativePath) {
+        return find(relativePath.split('/', QString::SkipEmptyParts));
+    }
+    FileInfo *createDir(const QString &relativePath) {
+        const QStringList pathComponents = relativePath.split('/', QString::SkipEmptyParts);
+        const QString childName = pathComponents.last();
+        FileInfo *parent = findInvalidatingEtags(pathComponents.mid(0, pathComponents.size() - 1));
+        Q_ASSERT(parent);
+        FileInfo &child = parent->children[childName] = FileInfo{childName};
+        child.parentPath = parent->path();
+        child.etag = QString::number(QDateTime::currentDateTime().toMSecsSinceEpoch(), 16);
+        return &child;
+    }
+    FileInfo *create(const QString &relativePath, qint64 size, char contentChar) {
+        const QStringList pathComponents = relativePath.split('/', QString::SkipEmptyParts);
+        const QString childName = pathComponents.last();
+        FileInfo *parent = findInvalidatingEtags(pathComponents.mid(0, pathComponents.size() - 1));
+        Q_ASSERT(parent);
+        FileInfo &child = parent->children[childName] = FileInfo{childName, size};
+        child.parentPath = parent->path();
+        child.contentChar = contentChar;
+        child.etag = QString::number(QDateTime::currentDateTime().toMSecsSinceEpoch(), 16);
+        return &child;
+    }
+    bool operator<(const FileInfo &other) const {
+        return name < other.name;
+    }
+
+    bool operator==(const FileInfo &other) const {
+        return name == other.name
+            && isDir == other.isDir
+            && size == other.size
+            && children == other.children;
```

Comment:
Added a comment, and a missing check for contentChar (and fix for the test failure that happened after adding that check).